python/ray/serve/pipeline/json_serde.py

```diff
@@ -0,0 +1,166 @@
+            # TODO: (jiaodong) Handle __main__ swap to filename
+            if isinstance(dag_node, ClassNode):
+                body = dag_node._body.__ray_actor_class__
+                import_path = f"{body.__module__}.{body.__qualname__}"
+            elif isinstance(dag_node, FunctionNode):
+                body = dag_node._body
+                import_path = f"{body.__module__}.{body.__qualname__}"
+            elif isinstance(dag_node, (DeploymentNode, DeploymentMethodNode)):
+                result_dict.update(
+                    {"deployment_name": dag_node._deployment.name}
+                )
+                if isinstance(dag_node._deployment._func_or_class, str):
+                    # We're processing a deserilized JSON node where import_path
+                    # is dag_node body.
+                    import_path = dag_node._deployment._func_or_class
+                else:
+                    body = (
+                        dag_node._deployment._func_or_class.__ray_actor_class__
+                    )
+                    import_path = f"{body.__module__}.{body.__qualname__}"
+                if isinstance(dag_node, DeploymentMethodNode):
+                    result_dict.update({"method_name": dag_node._method_name})
+
+            # TODO:(jiaodong) Maybe use cache for idential objects
+            result_dict.update(
+                # TODO: (jiaodong) Support runtime_env with remote working_dir
+                {"import_path": import_path}
+            )
+
+            return result_dict
+        else:
+            # Let the base class default method raise the TypeError
+            return json.JSONEncoder.default(self, obj)
+
+
+def dagnode_from_json(input_json: Any) -> Union[DAGNode, RayServeHandle, Any]:
+    """JSON serialization is only used and enforced in ray serve from ray core
+    API authored DAGNode(s).
+    """
+    # Deserialize a RayServeHandle.
+    if (
+        SERVE_SYNC_HANDLE_JSON_PREFIX_KEY in input_json
+        or SERVE_ASYNC_HANDLE_JSON_PREFIX_KEY in input_json
+    ):
+        return json.loads(input_json, object_hook=serve_handle_object_hook)
+    # Base case for plain objects.
+    elif DAGNODE_TYPE_KEY not in input_json:
+        return input_json
+
+    try:
+        # Check if input is JSON deserializable
+        return json.loads(input_json)
+    except Exception:
+        if input_json[DAGNODE_TYPE_KEY] == InputNode.__name__:
+            # TODO: (jiaodong) Support user passing inputs to InputNode in JSON
+            return InputNode()
+
+        # Pre-order JSON deserialization
+        args = json.loads(input_json["args"], object_hook=dagnode_from_json)
+        kwargs = json.loads(input_json["kwargs"], object_hook=dagnode_from_json)
+        # .options() should not contain any DAGNode type
+        options = json.loads(input_json["options"])
+        other_args_to_resolve = json.loads(
+            input_json["other_args_to_resolve"], object_hook=dagnode_from_json
+        )
```

Comment:
don't do all of this handling in the `except` block, it'll lead to nasty stack traces